### PR TITLE
New network Top level resource: VpnServerConfiguration related APIs

### PR DESCRIPTION
**Pull request by Nilambari**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by Nilambari** [comment]
Label: DoNotMerge. to add. I have opened this PR for ARM team to take a look and approve ARM manifest update for canary region for testing. We do not want to merge this PR yet.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 1]
```diff
@@ -2593,7 +2901,330 @@
...
+          "description": "AAD Vpn authentication parameter AAD tenant."
+        },
+        "aadAudience": {
+          "type": "string",
+          "description": "AAD Vpn authentication parameter AAD audience."
+        },
+        "aadIssuer": {
+          "type": "string",
+          "description": "AAD Vpn authentication parameter AAD issuer."
+        },
```
Comment (nschonni) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: trailing comma
```suggestion
        }
```

**Comment by number213** [comment]
Added "do not merge" label by @Nilambari's request

**Review by EvgenyAgafonchikov — CHANGES_REQUESTED** [review]
Please run  validator
`oav validate-example <path>/virtualWan.json`
and fix issues in the examples

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 1 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Done.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 2]
```diff
@@ -2593,7 +2901,330 @@
...
+              "AAD"			  
+            ],
+            "x-ms-enum": {
+              "name": "VpnAuthenticationType",
+              "modelAsString": true
+            }
+          },
+          "description": "VPN authentication types for the VpnServerConfiguration."
+        },		
+        "VpnServerConfigVpnClientRootCertificates": {
```
[the remark this answers is not in the document]
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Done.

**Comment by Nilambari** [comment]
> Please run  validator
> oav validate-example <path>/virtualWan.json
> and fix issues in the examples

Somehow oav tool is not working for me. Can you let me know what changes are needed.

**Comment by Nilambari** [comment]
> Please run  validator
> oav validate-example <path>/virtualWan.json
> and fix issues in the examples

@EvgenyAgafonchikov Somehow oav tool is not working for me. Can you let me know what changes are needed.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 3]
```diff
@@ -646,7 +646,315 @@
...
+            }
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },	
+	
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualWans/{virtualWANName}/vpnConfiguration": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: POST actions are expected to be named with a leading verb, such as:
`exportVPNConfiguration` or `downloadVPNConfiguration`

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 4]
```diff
@@ -646,7 +646,315 @@
...
+          },
+          {
+            "$ref": "./network.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "name": "VpnServerConfigurationParameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "./network.json#/definitions/TagsObject"
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: It's not blocking, but we generally prefer to see all mutable properties supported in PATCH, even though `tags` is the only required property for patch.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 5]
```diff
@@ -646,7 +646,315 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Request successful. Returns the details of the VpnServerConfiguration updated.",
+            "schema": {
+              "$ref": "#/definitions/VpnServerConfiguration"
+            }
+          },
+          "201": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Given that this only supports `tags` for PATCH, would it be an async operation?

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 6]
```diff
@@ -2593,7 +2901,330 @@
...
+          "type": "string",
+          "description": "The provisioning state of the VpnServerConfiguration VPN client root certificate resource. Possible values are: 'Updating', 'Deleting', and 'Failed'."
+        }
+      },
+      "required": [
+        "publicCertData"
+      ],
+      "description": "Properties of VPN client root certificate of VpnServerConfiguration."
+    },
+    "VpnServerConfigVpnClientRootCertificate": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: I'm not entirely sure why these certificates (which are properties of the vpnServerConfiguration resource type), are partially modeled as if they are an ARM resource. That is, they have an object schema like:
`"id":<>,
"name":<>,
"properties":{}
`
yet they aren't really ARM resources themselves. Do they need to be modeled this way? Would it caused confusion that the `id` isn't actually routable via ARM?

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 7]
```diff
@@ -2593,7 +2901,330 @@
...
+      ],
+      "description": "VPN client root certificate of VpnServerConfiguration."
+    },
+    "VpnServerConfigRadiusServerRootCertificatePropertiesFormat": {
+      "properties": {
+        "publicCertData": {
+          "type": "string",
+          "description": "The certificate public data."
+        },
+        "provisioningState": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: What is the function of `provisioningState` on the certificates? Is that used to indicate status of some kind of async instantiation? Does the `provisioningState` property of vpnServerConfigurations resource reflect these in aggregate?

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 8]
```diff
@@ -2593,7 +2901,330 @@
...
+          "description": "VPN client revoked certificate of VpnServerConfiguration."
+        },
+        "vpnServerConfigRadiusServerRootCertificates": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/VpnServerConfigRadiusServerRootCertificate"
+          },
+          "description": "Radius Server root certificate of VpnServerConfiguration."
+        },
+        "vpnServerConfigRadiusClientRootCertificates": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: A number of these certificate properties share a common schema, but each have a discrete definition. Will they change independently in the future? A common definition could be used.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 9]
```diff
@@ -2593,7 +2901,330 @@
...
+        "aadAuthenticationParameters": {
+          "$ref": "#/definitions/AadAuthenticationParameters",
+          "description": "The set of aad vpn authentication parameters."
+        },
+        "provisioningState": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The provisioning state of the VpnServerConfiguration resource. Possible values are: 'Updating', 'Deleting', and 'Failed'."
+        },
+        "p2SVpnGateways": {
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Friendly reminder to make sure linked access checks are defined in your manifest to ensure proper authorization.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json [hunk 10]
```diff
@@ -2593,7 +2901,330 @@
...
+        },
+        "etag": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Gets a unique read-only string that changes whenever the resource is updated."
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "./network.json#/definitions/SubResource"
```
Comment (KrisBash) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: I assume this should be a reference to: "./network.json#/definitions/Resource" and not "SubResource"?  Taking "SubResource" into account, this is missing: `type`, `tags`, and `location`.

**Review by KrisBash — CHANGES_REQUESTED** [review]
Please see comments

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 3 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Sorry for adding confusion. This api is not needed. Removed it.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 4 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: So this patch api is added only for updating tags, which is used for billing purpose internally. We do not want to allow to change any other properties in this api. User can use PUT api if they wish to update any properties.

Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Actually Patch api is also not needed right now, so no need to add in here right now. Will add it if needed in future.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 5 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Removed patch api.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 6 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: All 4 types of certificates which are properties of vpnServerConfiguration resource are child resource of it, so they will have all the schema like id, name, etc. Also, currently we have not exposed any rest apis to directly update these child resources but that's doable.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 7 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Provisioning state on the certificate shows the state of that certificate child resource. This is child resource of top level resource: vpnserverconfiguration so they have provisioning state as well in NRP.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 8 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Certificates:- vpnServerConfigVpnClientRootCertificates and vpnServerConfigRadiusServerRootCertificates share common scheama and certificates:- vpnServerConfigVpnClientRevokedCertificates and vpnServerConfigRadiusClientRootCertificates share common schema. But, they represent different definitions e.g one is related to Certifiate vpn authentication and others are related to Radius vpn authentication so we want to keep them seperate.

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 10 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Certificates resources e.g. VpnServerConfigVpnClientRootCertificate are child resources of top level resource: VpnServerConfiguration. Isn't this correct then?

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json (hunk 9 above)
Comment (Nilambari) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/virtualWan.json]: Sure, I will update maifest PR with linked access checks.

**Comment by Nilambari** [comment]
@KrisBash  Thanks a lot for taking a look at PR. Incorporated review comments. Please take a look.

**Comment by number213** [comment]
Please retarget to `master` once Network's release branch is merged (https://github.com/Azure/azure-rest-api-specs/pull/6632)

**Comment by Nilambari** [comment]
@KrisBash Can you let me know if changes look ok here. I will open new PR for this targeting to master branch later as suggested?

**Pull request closed without merging**